geomesa-fs/geomesa-fs-tools/src/main/scala/org/locationtech/geomesa/fs/tools/ingest/FsManageMetadataCommand.scala

```diff
@@ -82,6 +97,229 @@ object FsManageMetadataCommand {
+              Command.output.info(strings.sorted.mkString("  ", "\n  ", ""))
+            }
+          }
+        }
+      }
+
+      override def close(): Unit = pool.shutdown()
+
+      /**
+       * List the files in the root directory. Excludes known file-based metadata files. Results are
+       * added to onDisk
+       */
+      private def listRoot(): Unit = {
+        val iter = storage.context.fc.listStatus(storage.context.root)
+        // use a phaser to track worker thread completion
+        val phaser = new Phaser(2) // 1 for this thread + 1 for the worker
+        pool.submit(new TopLevelListWorker(phaser, iter))
+        // wait for the worker threads to complete
+        phaser.awaitAdvanceInterruptibly(phaser.arrive())
+      }
+
+      private def buildPartitionConfigs(): (Int, Seq[PartitionMetadata]) = {
+        var count = 0
+        val partitions = Seq.newBuilder[PartitionMetadata]
+        onDisk.asScala.foreach { case (partition, files) =>
+          val update = Seq.newBuilder[StorageFile]
+          files.asScala.foreach { case (name, _) =>
+            update += StorageFile(name, System.currentTimeMillis())
+            count += 1
+          }
+          partitions += PartitionMetadata(partition, update.result, None, 0L)
+        }
+        (count, partitions.result)
+      }
+
+      private class TopLevelListWorker(phaser: Phaser, list: RemoteIterator[FileStatus]) extends Runnable {
+        override def run(): Unit = {
+          try {
+            var i = phaser.getRegisteredParties + 1
+            while (list.hasNext && i < PhaserUtils.MaxParties) {
+              val status = list.next
+              val path = status.getPath
+              val name = path.getName
+              if (status.isDirectory) {
+                if (name != FileBasedMetadataFactory.MetadataDirectory &&
+                    partitions.forall(_.exists(p => p == name || p.startsWith(s"$name/")))) {
+                  i += 1
+                  // use a tiered phaser on each directory avoid the limit of 65535 registered parties
+                  pool.submit(new ListWorker(new Phaser(phaser, 1), name, storage.context.fc.listStatus(path)))
+                }
+              } else if (name != MetadataJson.MetadataPath) {
+                onDisk.computeIfAbsent("", computeFunction).put(name, java.lang.Boolean.TRUE)
+              }
+            }
+            if (list.hasNext) {
+              pool.submit(new TopLevelListWorker(new Phaser(phaser, 1), list))
+            }
+          } finally {
+            phaser.arriveAndDeregister()
+          }
+        }
+      }
+
+      private class ListWorker(phaser: Phaser, partition: String, listDirectory: => RemoteIterator[FileStatus])
+          extends Runnable {
+        override def run(): Unit = {
+          try {
+            var i = phaser.getRegisteredParties + 1
+            val iter = listDirectory
+            while (iter.hasNext && i < PhaserUtils.MaxParties) {
```

Comment:
Is this pattern for having multiple Phasers worth abstracting?

Reply:
possibly, but it didn't seem particularly easy to do... we're using it in 2 places, if we get a third we can abstract it then ;)